### ISSUE TITLE
Stack level too deep

### ISSUE THREAD

**Issue by vihai** [dmendel/bindata#93]
Hello,

I have a piece of code like the following. Of course the actual buffer content is more complex and do require reading until ::eof

```
class Test < BinData::Record
  endian :little

  uint8 :len, value: lambda { payload.num_bytes + payload.rel_offset }

  buffer :payload, length: lambda { len - payload.rel_offset } do
    array read_until: :eof do
      uint8 :entry
    end
  end
end
```

I get this error:

```
[25] pry(main)> Test.new(payload: [ 1, 2, 3, 4, 5 ]).inspect
SystemStackError: stack level too deep
from (pry):9:in `block in <class:Test>'

[26] pry(main)> wtf?!?
Exception: SystemStackError: stack level too deep
--
 0: (pry):9:in `block in <class:Test>'
 1: /var/lib/gems/2.3.0/gems/bindata-2.4.0/lib/bindata/lazy.rb:33:in `instance_exec'
 2: /var/lib/gems/2.3.0/gems/bindata-2.4.0/lib/bindata/lazy.rb:33:in `lazy_eval'
 3: /var/lib/gems/2.3.0/gems/bindata-2.4.0/lib/bindata/base.rb:113:in `eval_parameter'
 4: /var/lib/gems/2.3.0/gems/bindata-2.4.0/lib/bindata/buffer.rb:104:in `do_num_bytes'
 5: /var/lib/gems/2.3.0/gems/bindata-2.4.0/lib/bindata/base.rb:168:in `num_bytes'
 6: (pry):7:in `block in <class:Test>'
 7: /var/lib/gems/2.3.0/gems/bindata-2.4.0/lib/bindata/lazy.rb:33:in `instance_exec'
 8: /var/lib/gems/2.3.0/gems/bindata-2.4.0/lib/bindata/lazy.rb:33:in `lazy_eval'
 9: /var/lib/gems/2.3.0/gems/bindata-2.4.0/lib/bindata/base.rb:113:in `eval_parameter'
10: /var/lib/gems/2.3.0/gems/bindata-2.4.0/lib/bindata/base_primitive.rb:157:in `_value'
11: /var/lib/gems/2.3.0/gems/bindata-2.4.0/lib/bindata/base_primitive.rb:86:in `snapshot'
12: /var/lib/gems/2.3.0/gems/bindata-2.4.0/lib/bindata/base_primitive.rb:98:in `respond_to?'
13: /var/lib/gems/2.3.0/gems/bindata-2.4.0/lib/bindata/lazy.rb:98:in `recursively_eval'
14: /var/lib/gems/2.3.0/gems/bindata-2.4.0/lib/bindata/lazy.rb:80:in `eval_symbol_in_parent_context'
15: /var/lib/gems/2.3.0/gems/bindata-2.4.0/lib/bindata/lazy.rb:69:in `method_missing'
16: (pry):9:in `block in <class:Test>'
17: /var/lib/gems/2.3.0/gems/bindata-2.4.0/lib/bindata/lazy.rb:33:in `instance_exec'
18: /var/lib/gems/2.3.0/gems/bindata-2.4.0/lib/bindata/lazy.rb:33:in `lazy_eval'
19: /var/lib/gems/2.3.0/gems/bindata-2.4.0/lib/bindata/base.rb:113:in `eval_parameter'
20: /var/lib/gems/2.3.0/gems/bindata-2.4.0/lib/bindata/buffer.rb:104:in `do_num_bytes'
21: /var/lib/gems/2.3.0/gems/bindata-2.4.0/lib/bindata/base.rb:168:in `num_bytes'
22: (pry):7:in `block in <class:Test>'
23: /var/lib/gems/2.3.0/gems/bindata-2.4.0/lib/bindata/lazy.rb:33:in `instance_exec'
24: /var/lib/gems/2.3.0/gems/bindata-2.4.0/lib/bindata/lazy.rb:33:in `lazy_eval'
25: /var/lib/gems/2.3.0/gems/bindata-2.4.0/lib/bindata/base.rb:113:in `eval_parameter'
26: /var/lib/gems/2.3.0/gems/bindata-2.4.0/lib/bindata/base_primitive.rb:157:in `_value'
27: /var/lib/gems/2.3.0/gems/bindata-2.4.0/lib/bindata/base_primitive.rb:86:in `snapshot'
28: /var/lib/gems/2.3.0/gems/bindata-2.4.0/lib/bindata/base_primitive.rb:98:in `respond_to?'
29: /var/lib/gems/2.3.0/gems/bindata-2.4.0/lib/bindata/lazy.rb:98:in `recursively_eval'

```

What should be the proper way to implement such kind of structures?

**Comment by dmendel** [comment]
```
class Test < BinData::Record
  uint8 :len, value: lambda { payload.num_bytes }
  buffer :payload, length: lambda { len }
end
```
`:len` and `:payload` are mutually recursive, hence your error about stack level too deep.

> What should be the proper way to implement such kind of structures?

What are you trying to achieve?

**Comment by vihai** [comment]
I'm implementing the NetLink protocol. A netlink packet had a length, some fixed fields, a payload type ad a payload.

The payload itself has (generally) some fixed fields and a sequence of TLVs.

I have to stop reading the TLVs when there are no more available bytes in the payload. I was using a complex expression taking the payload size from the parent but it was failing when there were no TLVs.

I wanted to use the buffer approach to have a strict constraint on the payload size and read the payload until EOS.

That said, the declaration I pasted you is IMHO semantically legitimate as the "len" field and the length of the buffer are bound. One is calculated from the other when the packet is either read or built. I think there should be some sort of two-way binding (eventually splitting the decoding/encoding path).

**Comment by vihai** [comment]
I'm pasting a snippet of the actual code:

```
  class S_link_attr < BinData::Record
    endian native_endianness
    uint16 :rta_len, value: lambda { data.num_bytes + 4 }   # Length of option
    uint16 :rta_type  # Type of option

    choice :data, selection: :rta_type do
      ...
    end
  end

  class S_ifinfomsg < BinData::Record
    endian native_endianness
    uint8  :ifi_family # AF_UNSPEC
    uint8  :ifi_pad1
    uint16 :ifi_type   # Device type
    int32  :ifi_index  # Interface index
    uint32 :ifi_flags  # Device flags
    uint32 :ifi_change # change mask

    array :attrs, read_until: lambda { attrs.num_bytes >= (parent.nlmsg_len - parent.payload.rel_offset - attrs.rel_offset) }, onlyif: lambda { attrs.num_bytes < (parent.nlmsg_len - parent.payload.rel_offset - attrs.rel_offset) } do
      s_link_attr byte_align: 4
    end
  end

  class S_nlmsg < BinData::Record
    endian native_endianness
    uint32 :nlmsg_len, value: lambda { payload.num_bytes + payload.rel_offset }
    uint16 :nlmsg_type
    uint16 :nlmsg_flags
    uint32 :nlmsg_seq
    uint32 :nlmsg_pid

    choice :payload, selection: :nlmsg_type do
      s_nlmsg_error NLMSG_ERROR
      s_ifinfomsg RTM_GETLINK
      s_ifinfomsg RTM_NEWLINK
      string :default, read_length: lambda { nlmsg_len - payload.rel_offset }
    end
  end

```

If I remove onlyif it breaks due to eos while attempting to read an empty payload.
With onlyif: I get a stack overflow due to recursion between num_bytes and value

**Comment by dmendel** [comment]
Okay, I understand now. Use `Buffer#raw_num_bytes`.

```
class Test < BinData::Record
  endian :little

  uint8 :len, value: lambda { payload.raw_num_bytes + payload.rel_offset }

  buffer :payload, length: lambda { len - payload.rel_offset } do
    array read_until: :eof do
      uint8 :entry
    end
  end
end
```

**Comment by ccoenen** [comment]
If someone comes here via google: I had the same effect when trying to use the rel_offset of the very same field somewhat like this:

`string :alignment_padding, length: -> { (4 - (alignment_padding.rel_offset % 4)) % 4 }`

(was supposed to pad a given chuck of data to be a multiple of 4 bytes, there may be a more elegant way, this is just an illustration of what didn't work either.)

**Comment by dmendel** [comment]
Fixed in 2.4.1

As for the more elegant way, you may like to use [aligned fields](https://github.com/dmendel/bindata/wiki/Records#aligned-fields) if you don't care about the padding.
